tests/models/test_selector.py

```diff
@@ -0,0 +1,223 @@
+    """Given `model_rules` the ModelSelector should create the expected decision tree."""
+
+    model_selector = ModelSelector()
+    assert model_selector.get_decision_tree(model_rules) == decision_tree
+
+
+@pytest.mark.parametrize(
+    "model_rules,rules",
+    [
+        # rules are equal to ServerEventModel rules.
+        ({ServerEventModel: selector(foo="foo")}, selector(foo="foo")),
+        # rules are a subset of ServerEventModel rules.
+        ({ServerEventModel: selector(foo="foo", bar="bar")}, selector(bar="bar")),
+        # rules are a superset of ServerEventModel rules.
+        ({ServerEventModel: selector(bar="bar")}, selector(foo="foo", bar="bar")),
+    ],
+)
+def test_models_selector_model_selector_model_rules(model_rules, rules):
+    """Check that the ModelRules dictionary raises an exception when the provided
+    model has a list of rules that is a superset or subset of another list of rules.
+    """
+
+    model_rules = ModelRules(model_rules)
+    with pytest.raises(ModelRulesException):
+        model_rules[BaseModel] = rules
+
+
+@pytest.mark.parametrize(
+    "event,model",
+    [
+        (ServerEventFactory(), ServerEventModel),
+        (PageCloseBrowserEventFactory(), PageCloseBrowserEventModel),
+        ({"invalid": "event"}, None),
+    ],
+)
+def test_models_selector_model_selector_get_model(event, model):
+    """Given an event the ModelSelector.get_model method should return the corresponding model."""
+
+    event = json.loads(event.json()) if not isinstance(event, dict) else event
+    assert ModelSelector().get_model(event) is model
+
+
+@pytest.mark.parametrize(
+    "module,model_rules",
+    [
+        (type("empty_module", (), {}), {}),
+        (type("no_classes", (), {"int_member": 1, "str_member": "foo"}), {}),
+        (
+            type("not_a_subclass_of_base_model", (), {"not_base_model": ModelRules}),
+            {},
+        ),
+        (
+            type("missing_selector_attribute", (), {"no_selector": BaseModel}),
+            {},
+        ),
+        (
+            type(
+                "one_valid_base_model", (), {"page_close": PageCloseBrowserEventModel}
+            ),
+            {PageCloseBrowserEventModel: PageCloseBrowserEventModel.__selector__},
+        ),
+        (
+            type(
+                "two_valid_base_models_and_other_members_to_ignore",
+                (),
+                {
+                    "int_member": 1,
+                    "page_close": PageCloseBrowserEventModel,
+                    "no_selector": BaseModel,
+                    "not_base_model": ModelRules,
+                    "server": ServerEventModel,
+                    "str_member": "foo",
+                },
+            ),
+            {
+                PageCloseBrowserEventModel: PageCloseBrowserEventModel.__selector__,
+                ServerEventModel: ServerEventModel.__selector__,
+            },
+        ),
+    ],
+)
```

Comment:
Smart usage of `type` in a test!

![](https://media1.giphy.com/media/d3mlE7uhX8KFgEmY/giphy.gif)